cookbook/case_studies/ml_training/spark_horovod/keras_spark_rossmann_estimator.py

```diff
@@ -36,421 +38,608 @@
-                                         checkpoint_callback=ckpt_callback)
-
-    keras_model = keras_estimator.fit(train_df).setOutputCols(['Sales_output'])
+    backend = SparkBackend(
+        num_proc=hp.num_proc,
+        stdout=sys.stdout,
+        stderr=sys.stderr,
+        prefix_output_with_timestamp=True,
+    )
+    keras_estimator = hvd.KerasEstimator(
+        backend=backend,
+        store=store,
+        model=model,
+        optimizer=opt,
+        loss="mae",
+        metrics=[exp_rmspe],
+        custom_objects=CUSTOM_OBJECTS,
+        feature_cols=all_cols,
+        label_cols=["Sales"],
+        validation="Validation",
+        batch_size=hp.batch_size,
+        epochs=hp.epochs,
+        verbose=2,
+        checkpoint_callback=ckpt_callback,
+    )
 
-    history = keras_model.getHistory()
-    best_val_rmspe = min(history['val_exp_rmspe'])
-    print('Best RMSPE: %f' % best_val_rmspe)
+    keras_model = keras_estimator.fit(train_df).setOutputCols(["Sales_output"])
 
-    working_dir = flytekit.current_context().working_directory
+    history = keras_model.getHistory()
+    best_val_rmspe = min(history["val_exp_rmspe"])
+    print("Best RMSPE: %f" % best_val_rmspe)
 
     # Save the trained model.
     keras_model.save(os.path.join(working_dir, hp.local_checkpoint_file))
-    print('Written checkpoint to %s' % os.path.join(working_dir, hp.local_checkpoint_file))
+    print(
+        "Written checkpoint to %s" % os.path.join(working_dir, hp.local_checkpoint_file)
+    )
+    return keras_model
 
-    # ================ #
-    # FINAL PREDICTION #
-    # ================ #
 
-    print('================')
-    print('Final prediction')
-    print('================')
+def test(
+    keras_model,
+    working_dir: FlyteDirectory,
+    test_df: pyspark.sql.DataFrame,
+    hp: Hyperparameters,
+) -> FlyteDirectory:
 
-    pred_df=keras_model.transform(test_df)
+    print("================")
+    print("Final prediction")
+    print("================")
+
+    pred_df = keras_model.transform(test_df)
     pred_df.printSchema()
     pred_df.show(5)
 
-    # Convert from log domain to real Sales numbers
-    pred_df=pred_df.withColumn('Sales_pred', F.exp(pred_df.Sales_output))
+    # convert from log domain to real Sales numbers
+    pred_df = pred_df.withColumn("Sales_pred", F.exp(pred_df.Sales_output))
 
-    submission_df = pred_df.select(pred_df.Id.cast(T.IntegerType()), pred_df.Sales_pred).toPandas()
-    submission_df.sort_values(by=['Id']).to_csv(os.path.join(working_dir, hp.local_submission_csv), index=False)
-    print('Saved predictions to %s' % hp.local_submission_csv)
+    submission_df = pred_df.select(
+        pred_df.Id.cast(T.IntegerType()), pred_df.Sales_pred
+    ).toPandas()
+    submission_df.sort_values(by=["Id"]).to_csv(
+        os.path.join(working_dir, hp.local_submission_csv), index=False
+    )
+    print("Saved predictions to %s" % hp.local_submission_csv)
 
     return working_dir
 
 
+@task(
```

Comment:
lets put a comment block before this, else it breaks into a lot of code

Reply:
I have not added documentation at all. @SmritiSatyanV will be working on it later.

Reply:
ok